### PR TITLE
handle platform specific source path for json output

### LINKED ISSUE #3350: json match throws error due to source attribute mismatch on windows terminal
/kind bug

## What versions of software are you using?

**Operating System:**
win10

**Output of `odo version`:**
master

## How did you run odo exactly?

Ran `make test-cmd-cmp` on windows.

## Actual behavior

```
SSSSSCreated dir: C:\Users\Admin\AppData\Local\Temp\209745756
Creating a new project: xgvocbsngm
Running odo.exe with args [odo project create xgvocbsngm -w -v4]
[odo]  -  Waiting for project to come up  ...
[odo] I0615 05:53:34.877438    8988 occlient.go:543] Status of creation of project xgvocbsngm is Active
[odo] I0615 05:53:34.923391    8988 occlient.go:548] Project xgvocbsngm now exists
[odo] I0615 05:53:34.953368    8988 occlient.go:583] Status of creation of service account &ServiceAccount{ObjectMeta:{default  xgvocbsngm /api/v1/namespaces/xgvocbsngm/serviceaccounts/default 493ec3a9-5db9-4eae-ab6f-794c313ad5ed 42582 0 2020-06-15 05:53:35 +0000 GMT <nil> <nil> map[] map[] [] []  []},Secrets:[]ObjectReference{ObjectReference{Kind:,Namespace:,Name:default-token-mv29v,UID:,APIVersion:,ResourceVersion:,FieldPath:,},ObjectReference{Kind:,Namespace:,Name:default-dockercfg-dddjj,UID:,APIVersion:,ResourceVersion:,FieldPath:,},},ImagePullSecrets:[]LocalObjectReference{LocalObjectReference{Name:default-dockercfg-dddjj,},},AutomountServiceAccountToken:nil,} is ready
 V  Waiting for project to come up [1s]
[odo]  V  Project 'xgvocbsngm' is ready for use
[odo]  V  New project created and now using project: xgvocbsngm
[odo] I0615 05:53:34.974973    8988 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
Current working dir: C:\Users\Admin\go\src\github.com\openshift\odo\tests\integration
Setting current dir to: C:\Users\Admin\AppData\Local\Temp\209745756
Running odo.exe with args [odo component create nodejs cmp-git --project xgvocbsngm --context C:\Users\Admin\AppData\Local\Temp\209745756 --app testing -o json]
[odo] {
[odo]     "kind": "Component",
[odo]     "apiVersion": "odo.dev/v1alpha1",
[odo]     "metadata": {
[odo]         "name": "cmp-git",
[odo]         "namespace": "xgvocbsngm",
[odo]         "creationTimestamp": null
[odo]     },
[odo]     "spec": {
[odo]         "app": "testing",
[odo]         "type": "nodejs",
[odo]         "source": "file:///./",
[odo]         "sourceType": "local",
[odo]         "ports": [
[odo]             "8080/TCP"
[odo]         ]
[odo]     },
[odo]     "status": {
[odo]         "state": "Not Pushed"
[odo]     }
[odo] }
Deleting project: xgvocbsngm
Running odo.exe with args [odo project delete xgvocbsngm -f]
[odo]  V  Deleted project : xgvocbsngm
[odo]  !  Warning! Projects are deleted from the cluster asynchronously. Odo does its best to delete the project. Due to multi-tenant clusters, the project 
may still exist on a different node.
Setting current dir to: C:\Users\Admin\go\src\github.com\openshift\odo\tests\integration
Deleting dir: C:\Users\Admin\AppData\Local\Temp\209745756

------------------------------
+ Failure [2.093 seconds]
odo sub component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_sub_test.go:13
  Creating component
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:52
    should describe not pushed component when it is created with json output [It]
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:262

    Expected
        <string>: "...le:///./","..."
    to equal               |
        <string>: "...le://./","s..."

    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:268
```

## Expected behavior

json output should match all the attribute

## Any logs, error output, etc?

Manually verified on windows.

### PR DESCRIPTION
**Pull request by prietyc123**
**What type of PR is this?**

/kind bug

**What does does this PR do / why we need it**:

Fixing json mismatch on windows platform

**Which issue(s) this PR fixes**:

Fixes #3350 

**How to test changes / Special notes to the reviewer**:
`make test-cmd-cmp` and `make test-cmd-cmp-sub` should pass on windows

### CODE REVIEW

File: tests/integration/component.go [hunk 1]
```diff
@@ -263,7 +263,8 @@ func componentTests(args ...string) {
 			helper.CopyExample(filepath.Join("source", "nodejs"), context)
 			cmpDescribeJSON, err := helper.Unindented(helper.CmdShouldPass("odo", append(args, "create", "nodejs", "cmp-git", "--project", project, "--context", context, "--app", "testing", "-o", "json")...))
 			Expect(err).Should(BeNil())
-			expected, err := helper.Unindented(`{"kind": "Component","apiVersion": "odo.dev/v1alpha1","metadata": {"name": "cmp-git","namespace": "` + project + `","creationTimestamp": null},"spec":{"app": "testing","type":"nodejs","source": "file://./","sourceType": "local","ports": ["8080/TCP"]}, "status": {"state": "Not Pushed"}}`)
+			sourcePath := helper.PlatformSpecificSourcePath()
```
Comment (amitkrout) [inline tests/integration/component.go:266]: +1 for calling it through a helper function.

/lgtm

**Comment by prietyc123** [comment]
https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3357/pull-ci-openshift-odo-master-v4.3-integration-e2e/1273107433199243264#1:build-log.txt%3A2338

/retest

**Comment by prietyc123** [comment]
```
2020/06/17 16:15:29 Ran for 2h15m14s
error: some steps failed:
  * could not run steps: step integration-e2e failed: template pod "integration-e2e" failed: the pod ci-op-kr004k1v/integration-e2e failed after 2h5m47s (failed containers: test): ContainerFailed one or more containers exited
Container test exited with code 2, reason Error
```

/retest

**Comment by prietyc123** [comment]
Ran test on windows with current change

```
> git fetch origin pull/3357/head:pull_3357
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Total 7 (delta 5), reused 5 (delta 5), pack-reused 2
Unpacking objects: 100% (7/7), done.
From https://github.com/openshift/odo
 * [new ref]         refs/pull/3357/head -> pull_3357
 * [new tag]         v1.2.2     -> v1.2.2

> git checkout pull_3357
Switched to branch 'pull_3357'

> make test-cmd-cmp
ginkgo  -randomizeAllSpecs -slowSpecThreshold=120 -timeout 7200s -nodes=1 -focus="odo component command tests" tests/integration/
Running Suite: Integration Suite
================================
Random Seed: 1592470371 - Will randomize all specs
Will run 47 of 177 specs

S+++SSSSS+SSS++SSSSSSSS++SSSSSSSS+++SS++S
------------------------------
S [SKIPPING] in Spec Setup (JustBeforeEach) [0.002 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  Creating component using symlink
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:776
    Should be able to deploy a wildfly war file using symlinks in some odo commands [JustBeforeEach]
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:822

    Skipping test because for symlink creation on platform like Windows, go library needs elevated privileges.

    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:781
------------------------------
+S
------------------------------
+ [SLOW TEST:126.628 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  Creating component
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:52
    should list the component
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:194
------------------------------
SSSSrelativeContext = "..\\798730242"
+SSSS+S+++S+SSSSSSS++SS+SSSSS+S+SSSSkubeconfig before delete +SSSSSSSS+SSSSS+SSSSSSSS
------------------------------
S [SKIPPING] in Spec Setup (JustBeforeEach) [0.002 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  Creating component using symlink
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:776
    Should be able to deploy a spring boot uberjar file using symlinks in all odo commands [JustBeforeEach]
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:800

    Skipping test because for symlink creation on platform like Windows, go library needs elevated privileges.

    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:781
------------------------------
S+SSSS+SS+SSSSSSSSSSSS+SSS++S+S+SSS+SSSSSSS[.gitignore requirements.txt wsgi.py]
+S+SSS
------------------------------
+ [SLOW TEST:157.240 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  odo component updating
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:589
    should be able to create a git component and update it from local to git
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:598
------------------------------
+ [SLOW TEST:149.338 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  odo component updating
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:589
    should be able to update a component from git to local
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:615
------------------------------
SS
------------------------------
+ [SLOW TEST:120.669 seconds]
odo component command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  odo component delete should clean owned resources
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:841
    should delete the component and the owned resources
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/component.go:853
------------------------------
SSSSS+SS+SSSS
Ran 45 of 177 Specs in 2088.988 seconds
SUCCESS! -- 45 Passed | 0 Failed | 0 Pending | 132 Skipped
PASS

Ginkgo ran 1 suite in 34m53.6097736s
Test Suite Passed
```

**Comment by girishramnani** [comment]
/hold
There is an underlying issue we need to resolve on this

**Comment by amitkrout** [comment]
I have gone through the official Microsoft documentation [1] and Wikipedia doc [2] on file URI.

As per Wikipedia doc [2] on file URI : 
Over a network the file URI will be like file://localhost/<path> on both unix and windows setup and if the file is local then the URI will be file:///<path> on both platform

As per Microsoft documentation [1] on file URI : 
File URI will be like - file:///sDrives[|sFile]. whereas sDrives Specifies the local or network drive.

BTW, In our case we are dealing with the local component dir which reflects file URI in the json output against the source key. May be we need to revisit the backend code regarding how the json output prints file URI differently for the same local component on different platform

Lets not add bandage through the test pr. So i am closing the pr unti we get a proper fix. 

[1] https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/jj710207(v=vs.85)
[2] https://en.wikipedia.org/wiki/File_URI_scheme

Slack discussion - https://openshiftdo.slack.com/archives/CQK6HE78S/p1592466401238200

**Pull request closed without merging**